tornado/testing.py

```diff
@@ -499,13 +506,16 @@ def wrap(f):
         @functools.wraps(f)
         def pre_coroutine(self, *args, **kwargs):
             result = f(self, *args, **kwargs)
-            if isinstance(result, GeneratorType):
+            if isinstance(result, GeneratorType) or iscoroutine(result):
                 self._test_generator = result
             else:
                 self._test_generator = None
             return result
 
-        coro = gen.coroutine(pre_coroutine)
+        if iscoroutinefunction(f):
```

Comment:
If `@gen.coroutine` doesn't already support native coroutines in the way I had assumed, there's no need to change it. It never really makes sense to use `@gen.coroutine` on an `async def` function except that it allows (unusual) code like this to proceed without conditionals. Normally, the rule I would like to encourage is to use `gen.convert_yielded` on any non-None result (which guarantees a `Future`); we can't use that in `@gen_test` since it applies to generator functions without any other decorator. So I think this is fine as-is - I don't see any straightforward change that would improve it and it doesn't need the full generality of `convert_yielded`.